File: src/test/java/ru/jerael/booktracker/backend/data/repository/PublisherRepositoryImplTest.java
package ru.jerael.booktracker.backend.data.repository;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.data.jpa.test.autoconfigure.DataJpaTest;
import org.springframework.context.annotation.Import;
import ru.jerael.booktracker.backend.data.db.entity.PublisherEntity;
import ru.jerael.booktracker.backend.data.db.repository.JpaPublisherRepository;
import ru.jerael.booktracker.backend.data.mapper.PublisherDataMapper;
import ru.jerael.booktracker.backend.domain.model.publisher.Publisher;
import java.util.Optional;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.*;

@DataJpaTest
@Import({PublisherRepositoryImpl.class, PublisherDataMapper.class})
class PublisherRepositoryImplTest {

    @Autowired
    private PublisherRepositoryImpl publisherRepository;

    @Autowired
    private JpaPublisherRepository jpaPublisherRepository;

    private final String name = "Publisher Name";

    @Test
    void findByName_WhenExists_ShouldReturnPublisher() {
        PublisherEntity entity = new PublisherEntity();
        entity.setName(name);
        PublisherEntity savedEntity = jpaPublisherRepository.save(entity);

        Optional<Publisher> result = publisherRepository.findByName("publisher name");

        assertTrue(result.isPresent());
        assertEquals(savedEntity.getId(), result.get().id());
        assertEquals(name, result.get().name());
    }

    @Test
    void save_WhenIdIsNull_ShouldInsertNewPublisher() {
        Publisher publisher = new Publisher(null, name);

        Publisher result = publisherRepository.save(publisher);

        assertNotNull(result);
        assertEquals(name, result.name());
    }

    @Test
    void save_WhenIdIsPresent_ShouldUpdateExistingPublisher() {
        Publisher publisher = new Publisher(null, name);

        UUID savedId = publisherRepository.save(publisher).id();

        Publisher newPublisher = new Publisher(savedId, "New Publisher");

        Publisher result = publisherRepository.save(newPublisher);

        assertEquals(savedId, result.id());
        assertEquals("New Publisher", result.name());
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/data/repository/PublisherRepositoryImplTest.kt
package ru.jerael.booktracker.backend.data.repository

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertNotNull
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.data.jpa.test.autoconfigure.DataJpaTest
import org.springframework.context.annotation.Import
import ru.jerael.booktracker.backend.data.db.repository.JpaPublisherRepository
import ru.jerael.booktracker.backend.data.mapper.PublisherDataMapper
import ru.jerael.booktracker.backend.domain.model.pagination.PageQuery
import ru.jerael.booktracker.backend.domain.model.pagination.SortDirection
import ru.jerael.booktracker.backend.factory.publisher.PublisherDomainFactory
import ru.jerael.booktracker.backend.factory.publisher.PublisherEntityFactory

@DataJpaTest
@Import(PublisherRepositoryImpl::class, PublisherDataMapper::class)
class PublisherRepositoryImplTest {
    
    @Autowired
    private lateinit var jpaPublisherRepository: JpaPublisherRepository
    
    @Autowired
    private lateinit var publisherRepository: PublisherRepositoryImpl
    
    @Test
    fun `when publisher exists, findByName should return publisher`() {
        val publisher = jpaPublisherRepository.save(PublisherEntityFactory.createPublisherEntity())
        
        val result = publisherRepository.findByName(publisher.name)
        
        with(result.get()) {
            assertEquals(publisher.id, id)
            assertEquals(publisher.name, name)
        }
    }
    
    @Test
    fun `when id is null, save should insert new publisher`() {
        val publisher = PublisherDomainFactory.createPublisher(id = null)
        
        val result = publisherRepository.save(publisher)
        
        assertNotNull(result.id)
        assertEquals(publisher.name, result.name)
    }
    
    @Test
    fun `when id is present, save should update existing publisher`() {
        val savedPublisher = jpaPublisherRepository.save(PublisherEntityFactory.createPublisherEntity())
        val updatedPublisher = PublisherDomainFactory.createPublisher(id = savedPublisher.id, name = "New Publisher")
        
        val result = publisherRepository.save(updatedPublisher)
        
        assertEquals(savedPublisher.id, result.id)
        assertEquals(updatedPublisher.name, result.name)
    }
    
    @Test
    fun `searchByName should return list of found publishers by similarity`() {
        jpaPublisherRepository.saveAll(
            listOf(
                PublisherEntityFactory.createPublisherEntity(name = "Publisher A"),
                PublisherEntityFactory.createPublisherEntity(name = "Tor Books"),
                PublisherEntityFactory.createPublisherEntity(name = "Aboba Books")
            )
        )
        val query = "book"
        val pageQuery = PageQuery(0, 5, "name", SortDirection.ASC)
        
        val result = publisherRepository.searchByName(pageQuery, query)
        
        with(result) {
            assertThat(content).hasSize(2)
            
            val publisherNames = content.map { it.name }
            assertThat(publisherNames).containsExactlyInAnyOrder("Tor Books", "Aboba Books")
            assertThat(publisherNames).doesNotContain("Publisher A")
            
            assertThat(totalElements).isEqualTo(2)
        }
    }
}